<a href="https://colab.research.google.com/github/timurhatypov61-png/-2/blob/main/%D1%81%D1%82%D0%B0%D1%82%D0%B8%D1%81%D1%82%D0%B8%D0%BA%D0%B0%2C_Bootstrap.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np

delivery_time = np.array([35, 42, 28, 51, 39, 63, 31, 45])
order_price = np.array([1200, 3500, 800, 4200, 1900, 5600, 1500, 2800])
items_count = np.array([2, 5, 1, 7, 3, 8, 2, 4])

print(delivery_time.shape) #элементы по каждому измерению
print(order_price.shape)
print(items_count.shape)


print(delivery_time.ndim) #количество измерений
print(order_price.ndim)
print(items_count.ndim)

print(delivery_time.size) #общее кол элементов
print(order_price.size)
print(items_count.size)

(8,)
(8,)
(8,)
1
1
1
8
8
8


## Вывод по заданию 1.1

Все три массива одномерные (ndim = 1), содержат по 8 элементов (size = 8)
и имеют форму (8,).

Это означает, что данные согласованы: каждому из 8 заказов соответствует
ровно одно значение времени доставки, одна стоимость и одно количество товаров.

In [ ]:
print(delivery_time.size == order_price.size == items_count.size)

True


## Вывод по заданию 1.1

Все три массива одномерные (ndim = 1), содержат по 8 элементов (size = 8)
и имеют форму (8,). Проверка `delivery_time.size == order_price.size == items_count.size`
вернула True — значит, данные согласованы.

Каждый индекс в трёх массивах соответствует одному и тому же заказу:
delivery_time[0], order_price[0] и items_count[0] описывают один заказ №1,
delivery_time[1], order_price[1] и items_count[1] — заказ №2 и так далее.
Без этого соответствия нельзя было бы сопоставлять признаки между собой.

Если бы в одном массиве было 8 наблюдений, а в другом — 7, то Python
не смог бы корректно сопоставить данные, а с точки зрения качества данных
это означало бы потерю информации: для одного заказа неизвестен один из
признаков. Любой анализ, где нужны оба признака (например, средний чек по
времени доставки или корреляция), стал бы некорректным.

In [ ]:
orders = np.array([
    [35, 1200, 2],
    [42, 3500, 5],
    [28, 800, 1],
    [51, 4200, 7],
    [39, 1900, 3],
    [63, 5600, 8],
    [31, 1500, 2],
    [45, 2800, 4]
]) #1.время доставки; 2.стоимость заказа; 3.количество товаров.
print(orders.shape)

print(orders[0]) #данные первого заказа
print(orders[:,0]) #время доставки всех заказов
print(orders[:,1]) #стоимость всех заказов
print(orders[5,2]) #количество товаров в шестом заказе

print(orders[2:]) #все заказы начиная с третьего
print(orders[:5, :2]) #первые пять заказов и первые два признака

(8, 3)
[  35 1200    2]
[35 42 28 51 39 63 31 45]
[1200 3500  800 4200 1900 5600 1500 2800]
8
[[  28  800    1]
 [  51 4200    7]
 [  39 1900    3]
 [  63 5600    8]
 [  31 1500    2]
 [  45 2800    4]]
[[  35 1200]
 [  42 3500]
 [  28  800]
 [  51 4200]
 [  39 1900]]


## Вывод по заданию 1.2

Матрица orders имеет форму (8, 3): 8 строк — это 8 заказов,
3 столбца — время доставки, стоимость и количество товаров.

Индексация orders[i, j] позволяет получить конкретный элемент,
orders[i] — целую строку (заказ), orders[:, j] — целый столбец (признак).
Срезы orders[a:b, c:d] позволяют выбрать подмножество строк и столбцов.

orders.shape возвращает (8, 3), потому что в матрице 8 строк и 3 столбца —
это описывает структуру данных. orders.size возвращает 24, потому что
общее количество элементов = 8 × 3 = 24 — это описывает объём данных.

In [ ]:
orders = np.array([
    [35, 1200, 2],
    [42, 3500, 5],
    [28, 800, 1],
    [51, 4200, 7],
    [39, 1900, 3],
    [63, 5600, 8],
    [31, 1500, 2],
    [45, 2800, 4]
]) #1.время доставки; 2.стоимость заказа; 3.количество товаров.

delivery = orders[:, 0] #столбец времени доставки

delivery[0] = 1000 #Измените первый элемент delivery на 1000
print(orders) #Выведите orders


[[1000 1200    2]
 [  42 3500    5]
 [  28  800    1]
 [  51 4200    7]
 [  39 1900    3]
 [  63 5600    8]
 [  31 1500    2]
 [  45 2800    4]]


первая строка orders изменилась — вместо [35 1200 2] стало [1000 1200 2]. То есть изменение delivery повлияло на оригинал. Это и есть view.

In [ ]:
delivery[0] = 35 #Восстановите исходную матрицу orders
delivery_copy = orders[:, 0].copy() #Создайте независимую копию
delivery_copy[0] = 1000 #Измените первый элемент delivery_copy
print(orders) #Снова выведите orders

[[  35 1200    2]
 [  42 3500    5]
 [  28  800    1]
 [  51 4200    7]
 [  39 1900    3]
 [  63 5600    8]
 [  31 1500    2]
 [  45 2800    4]]


## Вывод по заданию 1.3

Срез orders[:, 0] без .copy() возвращает view — представление на исходные
данные. Изменение view меняет и оригинальную матрицу, потому что оба
объекта ссылаются на одну и ту же память.

Метод .copy() создаёт независимую копию. Изменения в копии не влияют
на исходный массив.

Разница важна при обработке реальных данных: случайное изменение view
может незаметно испортить исходный датасет. Например, аналитик выделил
столбец с ценами, чтобы временно их нормализовать, и случайно перезаписал
исходные значения. Дальше все отчёты строятся на искажённых данных, и
ошибку можно заметить не сразу.

In [ ]:
delivery_time = np.array([35, 42, 28, 51, 39, 63, 31, 45])

mask = delivery_time > 45 #булева маска для поиска заказов с временем доставки больше 45 минут
print(mask) #вывод маски
print(delivery_time[mask]) #результат применения маски
print(delivery_time[mask].size) #количество проблемных заказов
print(delivery_time[mask].mean()) #среднее время доставки для этой группы


[False False False  True False  True False False]
[51 63]
2
57.0


## Вывод по заданию 1.4

Булева маска delivery_time > 45 выделяет заказы, где доставка заняла
строго больше 45 минут. Таких заказов два: 51 и 63 минуты. Среднее
время по ним — 57 минут.

Разница между > 45 и >= 45: строгое неравенство не включает заказ с
ровно 45 минутами, нестрогое — включает. Если изменить граничное условие,
изменится и количество «проблемных» заказов, и бизнес-метрика (например,
доля опозданий). Это влияет на отчётность, премии курьеров и решения
по логистике.

In [ ]:
order_price = np.array([1200, 3500, 800, 4200, 1900, 5600, 1500, 2800])
items_count = np.array([2, 5, 1, 7, 3, 8, 2, 4])
delivery_time = np.array([35, 42, 28, 51, 39, 63, 31, 45])

mask1 = (order_price > 3000) & (items_count > 4) #булева маска заказ > 3000р и кол > 4
print(order_price[mask1]) #стоимость подходящих заказов
print(delivery_time[mask1]) #время доставки этих заказов
print(delivery_time[mask1].mean()) #среднее время доставки выбранной группы



[3500 4200 5600]
[42 51 63]
52.0


## Вывод по заданию 1.5

Единая маска (order_price > 3000) & (items_count > 4) выделяет заказы,
которые одновременно дорогие и крупные по количеству товаров. Таких
заказов три: на 3500, 4200 и 5600 рублей. Среднее время их доставки —
52 минуты.

Анализ комбинации признаков полезнее, чем анализ каждого по отдельности,
потому что позволяет найти сегменты, которые не видны при взгляде на один
показатель. Например, по отдельности дорогие заказы доставляются нормально,
и крупные заказы тоже. Но их пересечение может выявить проблему: склад
дольше собирает большой дорогой заказ, и доставка затягивается. Такой
сегмент требует отдельного внимания логистики.

In [ ]:
delivery_time = np.array([8, 9, 10, 11, 12])

print(delivery_time.mean()) #среднее арифметическое
print(np.median(delivery_time)) #медианa

delivery_time_corrupted = np.array([8, 9, 10, 11, 12, 1000]) #добавили выброс

print(delivery_time_corrupted.mean()) #среднее арифметическое
print(np.median(delivery_time_corrupted)) #медианa



10.0
10.0
175.0
10.5


## Вывод по заданию 2.1

Без выброса среднее и медиана совпали (10.0), потому что данные
симметричны. После добавления значения 1000 среднее выросло до 175.0,
а медиана — только до 10.5.

Среднее изменилось сильно, потому что оно учитывает каждое значение
и чувствительно к экстремальным точкам. Медиана изменилась слабо, потому
что она определяется положением середины отсортированного ряда, а выброс
стоит на краю и на середину не влияет.

При наличии редких экстремальных значений лучше использовать медиану —
она точнее описывает типичное время доставки. Среднее информативнее,
когда данные симметричны и без выбросов, например, температура воздуха
или зарплаты в узком диапазоне.

In [ ]:
np.random.seed(42)
delivery_simulated = np.random.normal(
    loc=30,
    scale=5,
    size=1000) #1000 наблюдений времени доставки


print(delivery_simulated.mean()) #среднее значение
print(delivery_simulated.std()) #стандартное отклонение
print(np.median(delivery_simulated)) #медиана

print("--------------------")

print(np.quantile(delivery_simulated, 0.50)) #50 процентиль
print(np.quantile(delivery_simulated, 0.90)) #90 процентиль
print(np.quantile(delivery_simulated, 0.95)) #95 процентиль
print(np.quantile(delivery_simulated, 0.99)) #99 процентиль


30.096660279111628
4.893631038736771
30.12650306117444
--------------------
30.12650306117444
36.52822600933569
38.38486264207573
41.57984750401471


## Вывод по заданию 2.2

Сгенерированная выборка из 1000 наблюдений имеет среднее ≈ 30.1,
стандартное отклонение ≈ 4.9, что соответствует параметрам генерации
(loc=30, scale=5).

Процентили: p50 = 30.13, p90 = 36.53, p95 = 38.38, p99 = 41.58.
p50 совпадает с медианой.

SLA «90% заказов за 40 минут или быстрее» выполняется: p90 = 36.53 < 40.

Продуктовый вывод: SLA выполняется, но длинный хвост распределения
контролируется не идеально — p99 = 41.58 мин, то есть примерно 1% заказов
доставляется дольше 40 минут. Для повышения качества стоит следить
не только за p90, но и за p95/p99.

In [ ]:
service_A = np.array([28, 29, 30, 31, 32])
service_B = np.array([15, 20, 30, 40, 45])

print(service_A.mean()) #среднее
print(service_B.mean())

print(np.median(service_A)) #медиана
print(np.median(service_B))

print(service_A.std()) #стандартное отклонение
print(service_B.std())

print(np.quantile(service_A, 0.90)) #90 процентиль
print(np.quantile(service_B, 0.90))

30.0
30.0
30.0
30.0
1.4142135623730951
11.40175425099138
31.6
43.0


## Вывод по заданию 2.3

У обеих служб одинаковое среднее (30.0) и медиана (30.0), но совершенно
разный разброс: стандартное отклонение A = 1.41, B = 11.40. p90 у A = 31.6,
у B = 43.0.

Одинаковые средние не означают одинаковое качество: служба A предсказуема
(все доставки 28–32 мин), служба B непредсказуема (от 15 до 45 мин).
Для доставки дорогой электроники я выбрал бы службу A, потому что
предсказуемость важнее скорости. Ключевые показатели для решения:
стандартное отклонение и p90.

In [ ]:
np.random.seed(42)
items = np.random.randint(1, 10, 100) #100 кол товаров
delivery_time_correlated = (20 + 3 * items + np.random.normal(0, 3, 100)) #время доставки

print(np.corrcoef(items, delivery_time_correlated)[0, 1]) #коэффициент корреляции Пирсона


0.9343293664033213


## Вывод по заданию 2.4

Коэффициент корреляции Пирсона между количеством товаров и временем
доставки: r ≈ 0.934. Это сильная положительная линейная связь: чем
больше товаров в заказе, тем дольше доставка.

Связь не идеальна (не 1.0), потому что в данные добавлен случайный шум —
другие факторы тоже влияют на время (пробки, погода, загрузка склада).

С точки зрения работы склада и доставки: крупный заказ дольше собирается,
проверяется и упаковывается; курьеру может понадобиться больше места
в машине или несколько ходок.

## Вывод по заданию 2.5

Высокая корреляция между продажами мороженого и несчастными случаями
на воде не доказывает причинно-следственную связь. Корреляция показывает
только то, что две величины меняются вместе, но не говорит, какая из них
является причиной.

Скрытый фактор, влияющий на оба показателя одновременно — это лето
(жаркая погода). Механизм: в жару люди покупают мороженое и одновременно
чаще идут купаться, где растёт риск несчастных случаев. Обе величины —
следствия одной общей причины.

Мой пример ложной корреляции: количество сыгранных матчей в Dota
и температура воздуха. Летом я играю чаще (больше свободного времени),
и летом температура выше. Возникает корреляция, будто Dota связана
с температурой. На самом деле общий скрытый фактор — лето: оно и даёт
больше свободного времени для игр, и повышает температуру. Прямой
причинно-следственной связи между Dota и температурой нет.

In [ ]:
x = np.array([28, 31, 29, 45, 32, 30])

print(x.mean()) #среднее значение исходной выборки

np.random.seed(42)
B = 5000
bootstrap_means = []

for i in range(B):
    sample = np.random.choice(x, size=len(x), replace=True) #новый массив из 6 чисел
    bootstrap_means.append(sample.mean()) #среднее этого массива

bootstrap_means = np.array(bootstrap_means) #переписываем в массив нампай

print(np.quantile(bootstrap_means, 0.025)) #2.5 процентиль
print(np.quantile(bootstrap_means, 0.975)) #97.5 процентиль

32.5
29.166666666666668
37.666666666666664


## Вывод по заданию 3.1

Среднее исходной выборки — 32.5 минуты. Bootstrap (B = 5000) дал
95%-й доверительный интервал [29.17; 37.67]. Интервал широкий, потому
что выборка маленькая (6 наблюдений) и содержит выброс 45, который
увеличивает неопределённость.

## Вопросы для анализа по заданию 3.1

1. Bootstrap использует выборку с возвращением, чтобы имитировать
случайность и получить разные виртуальные выборки. Без возвращения
все выборки были бы одинаковыми, и неопределённость оценить не удалось бы.

2. Выброс 45 расширяет распределение bootstrap-средних: когда 45 попадает
в выборку несколько раз, среднее растёт; когда не попадает — падает.
Из-за этого интервал становится шире и асимметричнее.

3. При очень маленькой выборке bootstrap неустойчив: каждая выборка
собирается из ограниченного набора чисел, и замена одного значения
сильно меняет результат.

4. Утверждение неверно. Истинное среднее фиксировано, а не случайно.
Корректная интерпретация: если строить такие интервалы многократно
на разных выборках, около 95% из них накроют истинное среднее.

## Вывод по заданию 3.2

H₀: μ_A = μ_B — среднее время доставки в двух группах одинаковое,
новый алгоритм не даёт эффекта.

H₁: μ_B < μ_A — новый алгоритм уменьшает среднее время доставки.

Ошибка I рода: мы отвергли H₀, хотя она верна. То есть решили, что
алгоритм улучшает доставку, а на самом деле улучшения нет. Это
«ложная тревога».

Ошибка II рода: мы не отвергли H₀, хотя она ложна. То есть решили,
что улучшения нет, а на самом деле оно есть. Это «пропущенный эффект».

Ошибка I рода соответствует ложной тревоге, ошибка II рода —
пропущенному эффекту.

К прямым напрасным расходам компании ведёт ошибка I рода. Если мы
по ошибке решим, что алгоритм работает, и внедрим его, мы будем платить
5 млн рублей в год за решение, которое ничего не улучшает. Ошибка II
рода не даёт прямых расходов: мы просто не внедряем алгоритм и
упускаем потенциальную выгоду, но деньги не теряем.

## Вывод по заданию 3.3

p-value — это вероятность увидеть такие или более экстремальные данные,
при условии что H₀ верна.

Утверждение «p-value = 0.001 означает вероятность истинности H₀ = 0.1%»
неверно по двум причинам. Во-первых, p-value не говорит о вероятности H₀ —
оно рассчитывается в предположении, что H₀ верна. Во-вторых, 0.001 — это
вероятность конкретных данных при H₀, а не вероятность самой гипотезы.

Результат статистически значим при α = 0.05, потому что 0.001 < 0.05.

В ситуации с 1 млн пользователей и эффектом 3 секунды: результат
статистически значим, но этого недостаточно для внедрения.
Статистическая значимость зависит от размера выборки, а практическая
значимость — от того, заметен ли эффект пользователю и оправдывает
ли он затраты.

Перед внедрением нужно учесть:
- размер эффекта (3 секунды — незаметно для пользователя);
- доверительный интервал;
- экономическую целесообразность (3 секунды не стоят 5 млн/год);
- риски;
- качество данных;
- сегментацию пользователей.

## Итоговый вывод

Ситуация: среднее время доставки уменьшилось, но разброс вырос,
p90 не изменился, p95 увеличился, p-value < 0.05.

Перед внедрением я бы задал следующие вопросы:

1. Качество данных: как собирались данные? Правильно ли проведён
   A/B-тест? Нет ли ошибок логирования?

2. Размер эффекта: на сколько именно уменьшилось среднее? В секундах
   или минутах? Это практическая разница?

3. Распределение: как выглядит распределение времени доставки?
   Нет ли групп, где стало хуже?

4. Выбросы: чем вызван рост разброса и p95? Появились ли новые
   экстремальные доставки? Это реальные заказы или аномалии?

5. Процентили: почему p90 не изменился, а p95 вырос? Что происходит
   в хвосте? Сколько заказов стало дольше 40 минут?

6. Статистическая значимость: насколько маленькое p-value?
   Достаточна ли мощность теста?

7. Практическая значимость: заметит ли пользователь улучшение?
   Оправдывает ли оно риски?

8. Бизнес-эффект: как изменение отразится на NPS, повторных заказах,
   отзывах? Экономия покрывает 5 млн рублей в год?

Итоговая рекомендация: не внедрять сразу. Улучшение среднего при
росте p95 и разброса означает, что часть пользователей стала получать
доставку дольше — это может ухудшить опыт, даже если среднее стало
лучше. Нужны дополнительные исследования: сегментация, анализ причин
роста хвоста, оценка экономического эффекта.